## This is for the EPIC chip row data processing
library("ChAMP")
library("ggtree")
library("minfi")
library("limma")
library("sva")
library(doParallel)

registerDoParallel()
myload = champ.load(directory = "/Users/Aoji.Xie/Desktop/DNA_Methylation/PD_blood_DNAm_LABV_20190429_EPIC",arraytype="EPIC")
dim(myload$beta)
'''
Filtering probes with a detection p-value above 0.01.
    Removing 7951 probes.
    If a large number of probes have been removed, ChAMP suggests you to identify potentially bad samples

  Filtering BeadCount Start
    Filtering probes with a beadcount <3 in at least 5% of samples.
    Removing 3380 probes

  Filtering NoCG Start
    Only Keep CpGs, removing 2953 probes from the analysis.

  Filtering SNPs Start
    Using general EPIC SNP list for filtering.
    Filtering probes with SNPs as identified in Zhous Nucleic Acids Research Paper 2016.
Removing 95485 probes from the analysis.

Filtering MultiHit Start
Filtering probes that align to multiple locations as identified in Nordlund et al
Removing 11 probes from the analysis.

Filtering XY Start
Filtering probes located on X,Y chromosome, removing 16541 probes from the analysis.

Updating PD file

Fixing Outliers Start
Replacing all value smaller/equal to 0 with smallest positive value.
Replacing all value greater/equal to 1 with largest value below 1..
[ Section 2: Filtering Done ]

All filterings are Done, now you have 739597 probes and 136 samples.

[<<<<< ChAMP.FILTER END >>>>>>]
[===========================]
[You may want to process champ.QC() next.]

[<<<<< ChAMP.LOAD END >>>>>>]
[===========================]
[You may want to process champ.QC() next.]
'''


####################################################################################################################################

## MINFI import RGset 
## use noob for the normalization

target <- read.metharray.sheet("/Users/Aoji.Xie/Desktop/DNA_Methylation/PD_blood_DNAm_LABV_20190429_EPIC")
RGset <- read.metharray.exp(target = target)
MSet.noob <- preprocessNoob(RGset)
####################################
test_1 <-  ratioConvert(MSet.noob, what = "both", keepCN = TRUE)

my_M <- getM(test_1)
my_M <- as.data.frame(my_M)
head(my_M)
M_t <- as.data.frame(t(my_M))


## change col name to subject_id
pd <- pData(test_1)
pd <- as.data.frame(pd)


pd$names <- rownames(pd)
pd_tmp <- pd[, c("names","Sample_Name")]

df_merge = merge(pd_tmp, M_t, by = 0)
rownames(df_merge) = df_merge$Sample_Name

df_merge$Row.names = NULL
df_merge$names = NULL
df_merge$Sample_Name = NULL


df_merge_t <- t(df_merge)
df_merge_t <- as.data.frame(df_merge_t)


my_M_final <- df_merge_t
my_M_final = my_M_final[,as.character(myload$pd$Sample_Name)]


####modified###

my_M_final_filtered <- my_M_final[rownames(myload$beta),]


#################### correct control by full model###################################################################
## import the cell portions generated by CIBERSORT
CIBERSORT = read.table("/Users/Aoji.Xie/Desktop/DNA_Methylation/CIBERSORT/combined_result.txt", sep = "\t", header = T)
dim(CIBERSORT)
sample_info_raw = read.table("/Users/Aoji.Xie/Desktop/DNA_Methylation/sample_info_raw.txt", header = TRUE, sep = "\t")
dim(sample_info_raw)


sample_info <- myload$pd
sample_info <- merge(sample_info, CIBERSORT, by = "Sample_Name")

covariates <- sample_info_raw[,c("Subject_ID", "Sex", "Age", "tobacco_100_in_life", "BMI")]

P0072rep <- covariates[110,]
P0072rep $Subject_ID = "P0072rep" 

P0042rep <- covariates[87,]
P0042rep$Subject_ID <- "P0042rep"

C0005rep <- covariates[2,]
C0005rep$Subject_ID <- "C0005rep"

C0148rep <- covariates[64,]
C0148rep$Subject_ID = "C0148rep"

covariates <- rbind(covariates,P0072rep,P0042rep,C0005rep,C0148rep)
colnames(covariates)[1] <- "Sample_Name"


sample_info_all_for_svd <- merge(sample_info, covariates, by = "Sample_Name")

rownames(sample_info_all_for_svd) <- sample_info_all_for_svd$Sample_Name
sample_info_all_for_svd <- sample_info_all_for_svd[unlist(colnames(my_M_final_filtered)),]
rownames(sample_info_all_for_svd) <- NULL

## run Champ.SVD() to check the batch effects
SVD_after_filter <- champ.SVD(my_M_final_filtered,pd=sample_info_all_for_svd)



edata <- as.matrix(my_M_final_filtered)
pheno <- sample_info_all_for_svd
rownames(pheno) <- pheno$Sample_Name
pheno <- AnnotatedDataFrame(pheno)



test <-  ExpressionSet(assayData = edata, phenoData = pheno)
dim(pData(test))

## correct batch effects by using combat
mod1 <- model.matrix(~ Sample_Group + Sex + CD8T + CD4T+ Neu+ Bcell + Mono +BMI + tobacco_100_in_life + Age , data=pData(test))
combat_edata_TEST <- ComBat(dat= edata , batch=pData(test)$Array, mod = mod1, par.prior=TRUE)
combat_edata_1_TEST <- ComBat(dat=combat_edata_TEST, batch=pData(test)$Slide, mod = mod1, par.prior=TRUE)

## check batch effects correction after correction 
SVD_after_correct <- champ.SVD(combat_edata_1_TEST,pd=sample_info_all_for_svd)

## conbine the replications 
methy_data_array_slide_combat_for_model <- combat_edata_1_TEST
methy_data_array_slide_combat_for_model <- data.frame(methy_data_array_slide_combat_for_model)
methy_data_array_slide_combat_for_model$P0072 <- (methy_data_array_slide_combat_for_model$P0072 + methy_data_array_slide_combat_for_model$P0072rep)/2
methy_data_array_slide_combat_for_model$P0042 <- (methy_data_array_slide_combat_for_model$P0042 + methy_data_array_slide_combat_for_model$P0042rep)/2
methy_data_array_slide_combat_for_model$C0005 <- (methy_data_array_slide_combat_for_model$C0005 + methy_data_array_slide_combat_for_model$C0005rep)/2
methy_data_array_slide_combat_for_model$C0148 <- (methy_data_array_slide_combat_for_model$C0148 + methy_data_array_slide_combat_for_model$C0148rep)/2

methy_data_array_slide_combat_for_model$P0072rep <- NULL
methy_data_array_slide_combat_for_model$C0005rep <- NULL
methy_data_array_slide_combat_for_model$C0148rep <- NULL
methy_data_array_slide_combat_for_model$P0042rep <- NULL

p_data <- sample_info_raw
p_data <- na.omit(p_data)


methy_data_array_slide_combat_for_model <-  methy_data_array_slide_combat_for_model[,as.character(p_data$Subject_ID)]

write.table(methy_data_array_slide_combat_for_model,"methy_data_array_slide_combat_for_model.txt", sep = "\t", quote = F, row.names = T)



